ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsLocator.java

```diff
@@ -68,28 +73,64 @@
   }
 
   @Override
-  public List<KsqlNode> locate(
-      final Struct key,
+  public List<KsqlLocation> locate(
+      final List<Struct> keys,
       final RoutingOptions routingOptions,
       final RoutingFilterFactory routingFilterFactory
   ) {
-    final KeyQueryMetadata metadata = kafkaStreams
-        .queryMetadataForKey(stateStoreName, key, keySerializer);
-
-    // Fail fast if Streams not ready. Let client handle it
-    if (metadata == KeyQueryMetadata.NOT_AVAILABLE) {
-      LOG.debug("KeyQueryMetadata not available for state store {} and key {}",
-                stateStoreName, key);
-      throw new MaterializationException(String.format(
-          "KeyQueryMetadata not available for state store %s and key %s", stateStoreName, key));
-    }
+    // Maintain request order for reproducibility by using a LinkedHashMap, even though it's
+    // not a guarantee of the API.
+    final Map<Integer, List<KsqlNode>> locationsByPartition = new LinkedHashMap<>();
+    final Map<Integer, Set<Struct>> keysByPartition = new HashMap<>();
+    final Set<Integer> filterPartitions = routingOptions.getPartitions();
```

Comment:
1. Why do you need to know the partitions here? Is it because you don't know which partitions are active and which standbys? If so, we can get this information with `KafkaStreams.allMetadata()`.  If we know what each partition is, then for a forwarded request, we know we need to access the standby partitions whereas for a non-forwarded only the active ones. This could avoid sending the partitions over the network causing overhead.

2. If you include the partitions in the request (and the host per list<partition>), I think we can avoid all this lookup and grouping per keys for forwarded requests. I know we were already doing this for a single key but that was cheap. Now, I feel we do unnecessary work for forwarded requests. What is the benefit of doing the grouping again for forwarded requests?

Reply:
I was thinking that you can be sure if it is a forwarded request, you need to access the standby partitions. And if you can know which partitions are active and which standby, then you can answer the question which partitions should be used to lookup keys. A request that is not forwarded always looksup active partitions whereas a forwarded request always looks up standby partitions.

Reply:
Discussed offline. The reason we need the partitions is that we need to know also which standby we want since they are ordered so we cannot rely only on the round number